### PR TITLE
Remote: Postpone the block waiting in `afterCommand` to `BlockWaitingModule`

### LINKED ISSUE #14576: `--bes_timeout` is ignored in Bazel 5
### Description of the problem / feature request:

`--bes_timeout` doesn't seem to be respected anymore in Bazel 5.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.
This will complete successfully and upload the full BES even though it shouldn't because the timeout is set to a very low 1ms:
```
$ git clone git@github.com:bazelbuild/rules_apple.git
$ cd rules_apple
$ env USE_BAZEL_VERSION=5.0.0rc4 bazelisk build --remote_cache=grpcs://remote.buildbuddy.io --bes_backend=grpcs://remote.buildbuddy.io --bes_results_url=https://app.buildbuddy.io/invocation/ //examples/iOS/StickersApp --bes_timeout=1ms --cpu=ios_x86_64
```

While this will fail as expected in Bazel 4.2.1:
```
$ env USE_BAZEL_VERSION=4.2.1 bazelisk build --remote_cache=grpcs://remote.buildbuddy.io --bes_backend=grpcs://remote.buildbuddy.io --bes_results_url=https://app.buildbuddy.io/invocation/ //examples/iOS/StickersApp --bes_timeout=1ms --cpu=ios_x86_64
ERROR: The Build Event Protocol upload timed out. com.google.common.util.concurrent.TimeoutFuture$TimeoutFutureException: Timed out: com.google.common.util.concurrent.Futures$NonCancellationPropagatingFuture@43553e48[status=PENDING, info=[delegate=[com.google.common.util.concurrent.SettableFuture@4d92c852[status=PENDING]]]]
```

### What operating system are you running Bazel on?

> Replace this line with your answer.

macOS 12.1

### PR DESCRIPTION
**Pull request by coeuvre**
When implementing async upload, we introduced a block waiting behaviour in `RemoteModule#afterCommand` so that uploads happened in the background can be waited before the whole build complete.

However, there are other block waiting code in other module's `afterCommand` method (e.g. BES module). Block waiting in remote module will prevent other modules' `afterCommand` from executing until it's completed. This causes issues like #14576.

This PR adds a new module `BlockWaitingModule` whose sole purpose is to accept tasks submitted by other modules in `afterCommand` and block waiting all the tasks in its own `afterCommand` method. So those tasks can be executed in parallel.

This PR only updates RemoteModule's `afterCommand` method to submit block waiting task. Other modules should be updated to use `BlockWaitingModule` as well but that's beyond the scope this this PR.

This PR along with 73a76a8af802f721f9de0f01026c50c4e72b2d35 fix #14576.

### CODE REVIEW

**Comment by brentleyjones** [comment]
Does this also fix the fact that `--bes_upload_mode=async` stopped working?

**Comment by coeuvre** [comment]
My gut feeling is yes but do we have an issue already where I can learn more details from?

**Comment by brentleyjones** [comment]
https://github.com/bazelbuild/bazel/issues/14620

**Comment by coeuvre** [comment]
Well, this PR won't fix #14620 but I have another fix which can fix both #14576 and #14620. Closing this for now.

File: src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package com.google.devtools.build.lib.runtime;
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java:1]: nit: please add copyright header

File: src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.devtools.build.lib.util.AbruptExitException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import javax.annotation.Nullable;
+
+/** A {@link BlazeModule} that await for submitted tasks to terminate after every command. */
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java:13]: nit: await -> waits

**Review by meisterT — APPROVED** [review]

**Comment by brentleyjones** [comment]
@Wyverald This fixes a 5.0 regression. Can we add it to the 5.0.1 milestone?

**Comment by Wyverald** [comment]
The linked issue is already on the milestone :)

File: src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java (hunk 1 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java:14]: Done.

File: src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java (hunk 2 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java]: Done.

File: src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2021 The Bazel Authors. All rights reserved.
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java:1]: super-nit: we have 2022 already :-D

**Review by meisterT — APPROVED** [review]

File: src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java (hunk 3 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/runtime/BlockWaitingModule.java]: yyyyyyyyes!

**Pull request closed without merging**